### PR TITLE
Feature/build error type response

### PR DESCRIPTION
**Pull request by Wellingtonbsjr**
# Summary

Create method _WithRuleType_ in validation, to override property name with Prefix and Type.

**How to use method _WithRuleType_ in validation?**
```
RuleFor(x => x.Text)
                .NotEmpty()
                .WithMessage("Text can't be null")
                .WithRuleType("SOME_RULE");
```

**How property name stay?** 
```
{Prefix}.{RuleType}.{PropertyName}
_RuleType.SOME_RULE.Text
```
**Why use it?**
To identify Rules in InvalidModelStateResponseFactory and customize the HTTP response.
```
public static class MvcBuilderExtensions
{
    private const string PropertyDefault = "msg";

    /// <summary>
    /// Configure the custom bad request response format for all validation errors (model binding and ModelValidationFilter).
    /// </summary>
    /// <param name="builder"></param>
    /// <returns></returns>
    public static IMvcBuilder ConfigureValidationResponseFormat(this IMvcBuilder builder)
    {
        builder.ConfigureApiBehaviorOptions(options =>
        {
            options.InvalidModelStateResponseFactory = context =>
            {
                var response = new ErrorResponse
                {
                    Error = new Dictionary<string, string[]>(),
                    Type = GetRuleType(context)
                };

                foreach (var (key, value) in context.ModelState.Where(x => x.Value.Errors.Any()))
                {
                    var errorMessage = value.Errors.Select(e => e.ErrorMessage).ToArray();
                    var errorKey = key.Contains(RuleTypeConst.Prefix) ? PropertyDefault : key;
                    response.Error.Add(errorKey, errorMessage);
                }

                return new BadRequestObjectResult(response);
            };
        });

        return builder;
    }

    private static string GetRuleType(ActionContext context)
    {
        var type = "ERROR_DEFAULT";

        var errorType = context.ModelState.LastOrDefault(error => error.Key.Contains(RuleTypeConst.Prefix));

        if (!string.IsNullOrEmpty(errorType.Key))
            type = GetRuleType(errorType.Key);

        return type;
    }

    private static string GetRuleType(string key) =>
        key.Split('.')[1];
}
```

### CODE REVIEW

**Review by isaacnborges-ac — APPROVED** [review]

**Review by nunorelvao — APPROVED** [review]

File: src/RuleTypeConst.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+{
+    /// <summary>
+    /// Const used do configure method WithRuleType.
+    /// </summary>
+    public static class RuleTypeConst
+    {
+        /// <summary>
+        /// Prefix used to override property name.
+        /// </summary>
+        public const string Prefix = "_RuleType";
```
Comment (luizhlelis) [inline src/RuleTypeConst.cs:11]: **question:** maybe I'm missing something here, but is this prefix really required?

File: src/RuleTypeBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public static IRuleBuilderOptions<T, TProperty> WithRuleType<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, string type)
+        {
+            string propertyName = "";
+
+            var options = rule.Configure(x => propertyName = x.GetDisplayName(null));
+
+            if (!string.IsNullOrEmpty(propertyName))
+                options.WithName(propertyName);
+
+            options.WithErrorCode($"{RuleTypeConst.Prefix}.{type}");
```
Comment (luizhlelis) [inline src/RuleTypeBuilderExtensions.cs:29]: thought: we're gonna need to use the same extension method `WithErrorCode` in #8, but I think we can use it in both scenarios. We can set some predefined words as ErrorCodes to change the response status code.

File: src/RuleTypeConst.cs (hunk 1 above)
Comment (Wellingtonbsjr) [reply src/RuleTypeConst.cs:11]: @luizhlelis this prefix is important to set correctly type in response.
We only set type, if exists an internal ErrorCode with this prefix.

File: src/RuleTypeBuilderExtensions.cs (hunk 2 above)
Comment (Wellingtonbsjr) [reply src/RuleTypeBuilderExtensions.cs:29]: Nice @luizhlelis, make sense.
We can use a new Enum.ErrorsStatusCode in this lib, creating a new method like "WithStatusCode(Enum.ErrorsStatusCode)".

**Comment by Wellingtonbsjr** [comment]
We create other branch feature/customer-main, to follow this:
1 - Change lib to response Bad Request with the format defined in the ower [playbook](https://votorantimindustrial.sharepoint.com/:w:/r/sites/juntos_somos_mais/_layouts/15/Doc.aspx?sourcedoc=%7BE8895835-9397-4E19-9046-26D7168A931A%7D&file=Padr%C3%A3o%20de%20retorno%20das%20APIs.docx&action=default&mobileredirect=true).

```
{
    "error": {
        "Cnpj": [
            "O CNPJ do cliente deve ter 14 caracteres"
        ]
    },
    "type": "VALIDATION_ERRORS"
}
```
2 - Apply branch feature/customer-main in customer project;
3 - Later apply these changes to the branch main.

**Pull request closed without merging**